### PR TITLE
Update dcc-validation-rule.json

### PR DESCRIPTION
**Pull request by ioangut**
## Description

Enhance distribution service with Booster Notification feature.

### CODE REVIEW

File: services/distribution/src/main/java/app/coronawarn/server/services/distribution/assembly/component/CwaApiStructureProvider.java [hunk 1]
```diff
@@ -86,6 +91,7 @@
       versionDirectory.addWritableToAll(ignoredValue -> Optional.ofNullable(archive));
     });
 
+
```
Comment (FelixRottler) [inline services/distribution/src/main/java/app/coronawarn/server/services/distribution/assembly/component/CwaApiStructureProvider.java:94]:
```suggestion
```

File: services/distribution/src/main/java/app/coronawarn/server/services/distribution/dgc/DigitalGreenCertificateToCborMapping.java [hunk 2]
```diff
@@ -49,7 +49,12 @@ public DigitalGreenCertificateToCborMapping(DigitalCovidCertificateClient digita
    */
   public List<BusinessRule> constructRules(RuleType ruleType)
       throws DigitalCovidCertificateException, FetchBusinessRulesException {
-    List<BusinessRuleItem> businessRulesItems = digitalCovidCertificateClient.getRules();
+    List<BusinessRuleItem> businessRulesItems;
+    if (ruleType.name().equalsIgnoreCase(RuleType.BoosterNotification.name())) {
```
Comment (FelixRottler) [inline services/distribution/src/main/java/app/coronawarn/server/services/distribution/dgc/DigitalGreenCertificateToCborMapping.java:53]: Maybe we can use the enum to store the corresponding logic behind this in order to replace the if statement? For me the if-statement feels wrong here

File: services/distribution/src/main/java/app/coronawarn/server/services/distribution/dgc/client/ProdDigitalCovidCertificateClient.java [hunk 3]
```diff
@@ -67,12 +67,20 @@ public ValueSet getValueSet(String hash) throws FetchValueSetsException {
...
+  public List<BusinessRuleItem> getBnRules() throws FetchBusinessRulesException {
+    return getResponseAndTreatExceptions(digitalCovidCertificateClient::getBnRules,
+        "business rules",
+        FetchBusinessRulesException::new);
+  }
+
   @Override
   public BusinessRule getCountryRuleByHash(String country, String hash) throws FetchBusinessRulesException {
-    return getResponseAndTreatExceptions(
+    BusinessRule businessRule =  getResponseAndTreatExceptions(
```
Comment (FelixRottler) [inline services/distribution/src/main/java/app/coronawarn/server/services/distribution/dgc/client/ProdDigitalCovidCertificateClient.java:79]: Can we directly return here?

File: services/distribution/src/main/resources/application.yaml [hunk 4]
```diff
@@ -225,26 +225,27 @@ services:
       dgc-directory: ${DIGITAL_GREEN_CERTIFICATE_DIRECTORY:ehn-dgc}
       valuesets-file-name: ${DIGITAL_GREEN_CERTIFICATE_VALUESETS_FILE_NAME:value-sets}
       supported-languages: ${DIGITAL_GREEN_CERTIFICATE_SUPPORTED_LANGUAGES:DE, EN, BG, PL, RO, TR}
+      booster-notification: ${BOOSTER_NOTIFICATION_DIRECTORY:booster-notification-rules}
       dsc-client:
-        public-key: ${DSC_PUBLIC_KEY:MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAETHfi8foQF4UtSNVxSFxeu7W+gMxdSGElhdo7825SD3Lyb+Sqh4G6Kra0ro1BdrM6Qx+hsUx4Qwdby7QY0pzxyA==}
-        base-url: ${DSC_BASE_PATH:https://de.dscg.ubirch.com}
+        public-key: ${DSC_PUBLIC_KEY:MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEIxHvrv8jQx9OEzTZbsx1prQVQn/3ex0gMYf6GyaNBW0QKLMjrSDeN6HwSPM0QzhvhmyQUixl6l88A7Zpu5OWSw==}
```
Comment (FelixRottler) [inline services/distribution/src/main/resources/application.yaml:230]: What is the background of this change?

File: services/distribution/src/main/resources/application.yaml [hunk 5]
```diff
@@ -225,26 +225,27 @@ services:
...
-        # Can be found e.g. at https://distribution.dcc-rules.de/publickey
         public-key: ${DCC_PUBLIC_KEY:MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEnQmn1s+sTWqFPuRhqFRide7HD52SrsD01ZdX4cxN1r0pnZ0Zh7JNwZxmXlOQW5NDVlPdHq2QcD06gmP/10QizA==}
-        retry-period: 5
-        max-retry-period: 30
-        max-retry-attempts: 5
-        base-url: ${BUSSINESS_RULES_BASE_PATH:https://distribution.dcc-rules.de}
+        retry-period: 2
+        max-retry-period: 5
+        max-retry-attempts: 2
+        base-url: ${BUSSINESS_RULES_BASE_PATH:https://distribution-dfe4f5c711db.dcc-rules.de/}
```
Comment (FelixRottler) [inline services/distribution/src/main/resources/application.yaml:241]: Why was this changed?

Comment (FelixRottler) [reply services/distribution/src/main/resources/application.yaml:241]: Was it wrong previously?

File: services/distribution/src/main/resources/application.yaml [hunk 6]
```diff
@@ -225,26 +225,27 @@ services:
...
         country-list-path: ${BUSSINESS_RULES_COUNTRY_LIST_PATH:/countrylist}
         value-sets-path: ${BUSSINESS_RULES_VALUE_SETS_PATH:/valuesets}
         rules-path: ${BUSSINESS_RULES_RULES_PATH:/rules}
+        bn-rules-path: ${BUSSINESS_RULES_BN_RULES_PATH:/rules}
         ssl:
-          trust-store: ${DCC_TRUST_STORE:../../docker-compose-test-secrets/dcc_truststore}
-          trust-store-password: ${DCC_TRUST_STORE_PASSWORD:123456}
+          trust-store: ${DSC_TRUST_STORE:../../docker-compose-test-secrets/dsc_truststore}
```
Comment (FelixRottler) [inline services/distribution/src/main/resources/application.yaml:247]: Change from DCC to DSC? Is this correct?

File: services/distribution/src/test/java/app/coronawarn/server/services/distribution/assembly/component/BoosterNotificationStructureProviderTest.java [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    // create a specific test folder for later assertions of structures.
+    testOutputFolder.create();
+    File outputDirectory = testOutputFolder.newFolder(PARENT_TEST_FOLDER);
+    Directory<WritableOnDisk> testDirectory = new DirectoryOnDisk(outputDirectory);
+    when(outputDirectoryProvider.getDirectory()).thenReturn(testDirectory);
+  }
+
+  @Test
+  void should_contain_booster_notification_rules() throws FetchBusinessRulesException {
+    when(digitalCovidCertificateClient.getBnRules()).thenThrow(FetchBusinessRulesException.class);
```
Comment (FelixRottler) [inline services/distribution/src/test/java/app/coronawarn/server/services/distribution/assembly/component/BoosterNotificationStructureProviderTest.java:77]: Shouldnt we check also the files and the content?

File: services/distribution/src/test/java/app/coronawarn/server/services/distribution/assembly/component/BoosterNotificationStructureProviderTest.java [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+
+    Writable<WritableOnDisk> boosterNotification = getBoosterNotificationStructure();
+
+    assertThat(boosterNotification.getName()).isEqualTo("booster-notification-rules");
+  }
+
+  private Writable<WritableOnDisk> getBoosterNotificationStructure() {
+    BoosterNotificationStructureProvider underTest = new BoosterNotificationStructureProvider(
```
Comment (FelixRottler) [inline services/distribution/src/test/java/app/coronawarn/server/services/distribution/assembly/component/BoosterNotificationStructureProviderTest.java:85]: Is there any reason why this is constructed manually?  Usually we can also inject beans here, which makes it easier to see which class is tested

File: services/distribution/src/test/resources/application.yaml [hunk 9]
```diff
@@ -197,7 +198,8 @@ services:
         base-url: ${BUSSINESS_RULES_BASE_PATH:https://distribution-dfe4f5c711db.dcc-rules.de/}
         country-list-path: ${BUSSINESS_RULES_COUNTRY_LIST_PATH:/countrylist}
         value-sets-path: ${BUSSINESS_RULES_VALUE_SETS_PATH:/valuesets}
-        rules-path: ${BUSSINESS_RULES_RULES_PATH:/rules}
+        rules-path: ${BUSSINESS_RULES_PATH:/rules}
```
Comment (FelixRottler) [inline services/distribution/src/test/resources/application.yaml:201]: In the src/application.yml a lot of changes were made but not so many in the test/application.yaml, are there any missing here?

File: services/distribution/src/main/resources/application.yaml (hunk 4 above)
Comment (ioangut) [reply services/distribution/src/main/resources/application.yaml:230]: This was before used to connect to the PRD dsc client from the local environment. But since we had the problem in PRD to connect to the dsc because of the whitelisting IPs, we have the same problem from the local env. So we changed this to the dev landscape so we can run the distribution against dev landscape from DSC

File: services/distribution/src/main/java/app/coronawarn/server/services/distribution/dgc/client/ProdDigitalCovidCertificateClient.java (hunk 3 above)
Comment (ioangut) [reply services/distribution/src/main/java/app/coronawarn/server/services/distribution/dgc/client/ProdDigitalCovidCertificateClient.java:79]: was used for debugging, need to revert, thanks

File: services/distribution/src/test/resources/application.yaml (hunk 9 above)
Comment (ioangut) [reply services/distribution/src/test/resources/application.yaml:201]: I will adjust the main application.yaml

File: services/distribution/src/main/java/app/coronawarn/server/services/distribution/assembly/component/BoosterNotificationStructureProvider.java [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+ * Reads Booster notification rules from the respective files in the class path or from dcc client and build a {@link
+ * BoosterNotificationStructureProvider} with them.
+ */
+
+@Component
+public class BoosterNotificationStructureProvider {
+
+  private static final Logger logger = LoggerFactory.getLogger(BoosterNotificationStructureProvider.class);
+
+  public static final String EXPORT_BIN = "export.bin";
```
Comment (roesslerj) [inline services/distribution/src/main/java/app/coronawarn/server/services/distribution/assembly/component/BoosterNotificationStructureProvider.java:31]: Could we have a semantic name, such as EXPORT_BINARY_FILENAME or similar? As is, if the string changes, this needs to change as well...

**Review by roesslerj — CHANGES_REQUESTED** [review]

File: services/distribution/src/main/resources/application.yaml (hunk 5 above)
Comment (ioangut) [reply services/distribution/src/main/resources/application.yaml:241]: same as public-key

File: services/distribution/src/main/java/app/coronawarn/server/services/distribution/dgc/DigitalGreenCertificateToCborMapping.java (hunk 2 above)
Comment (ioangut) [reply services/distribution/src/main/java/app/coronawarn/server/services/distribution/dgc/DigitalGreenCertificateToCborMapping.java]: We tried the enum approach and did not work.

File: services/distribution/src/main/resources/application.yaml (hunk 6 above)
Comment (ioangut) [reply services/distribution/src/main/resources/application.yaml]: Copy-paste mistake, Thanks!

**Review by FelixRottler — APPROVED** [review]

File: services/distribution/src/main/java/app/coronawarn/server/services/distribution/assembly/component/BoosterNotificationStructureProvider.java [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+  /**
+   * Create business rules Archive. If any exception is thrown during fetching data and packaging process, an empty
+   * Archive will be published in order to not override any previous archive on CDN with broken data. Provided rules are
+   * filtered by rule type parameter which could be 'Acceptance', 'Invalidation' or 'BoosterNotification'.
+   *
+   * @param ruleType    - rule type to receive rules for
+   * @param archiveName - archive name for packaging rules
+   * @return - business rules archive
+   */
+  private Writable<WritableOnDisk> getBnRulesArchive(RuleType ruleType, String archiveName) {
```
Comment (alexmoroianu) [inline services/distribution/src/main/java/app/coronawarn/server/services/distribution/assembly/component/BoosterNotificationStructureProvider.java:61]: I would go with full BoosterNotification name in here also in order to keep consistency with all the namings in this class. Also for clarity.

File: services/distribution/src/main/java/app/coronawarn/server/services/distribution/dgc/client/DigitalCovidCertificateClient.java [hunk 12]
```diff
@@ -24,5 +24,7 @@
 
   List<BusinessRuleItem> getRules() throws FetchBusinessRulesException;
 
+  List<BusinessRuleItem> getBnRules() throws FetchBusinessRulesException;
```
Comment (alexmoroianu) [inline services/distribution/src/main/java/app/coronawarn/server/services/distribution/dgc/client/DigitalCovidCertificateClient.java:27]: same here. There are no other abbreviations in the namings of other methods.

File: services/distribution/src/main/java/app/coronawarn/server/services/distribution/dgc/client/DigitalCovidCertificateFeignClient.java [hunk 13]
```diff
@@ -52,12 +52,19 @@
   @GetMapping(value = "${services.distribution.digital-green-certificate.client.rules-path}")
   ResponseEntity<List<BusinessRuleItem>> getRules();
 
+  /**
+   * HTTP GET to return all business rules.
+   */
+  @Timed
+  @GetMapping(value = "${services.distribution.digital-green-certificate.client.bn-rules-path}")
+  ResponseEntity<List<BusinessRuleItem>> getBnRules();
```
Comment (alexmoroianu) [inline services/distribution/src/main/java/app/coronawarn/server/services/distribution/dgc/client/DigitalCovidCertificateFeignClient.java:60]: same about naming. also javadoc should be adapted accordingly.

File: services/distribution/src/main/java/app/coronawarn/server/services/distribution/dgc/client/ProdDigitalCovidCertificateClient.java [hunk 14]
```diff
@@ -67,6 +67,13 @@ public ValueSet getValueSet(String hash) throws FetchValueSetsException {
         FetchBusinessRulesException::new);
   }
 
+  @Override
+  public List<BusinessRuleItem> getBnRules() throws FetchBusinessRulesException {
+    return getResponseAndTreatExceptions(digitalCovidCertificateClient::getBnRules,
+        "business rules",
```
Comment (alexmoroianu) [inline services/distribution/src/main/java/app/coronawarn/server/services/distribution/dgc/client/ProdDigitalCovidCertificateClient.java:73]: the string argument here is used to log exceptions about failed calls. There are also another method using "business rules" entity name. Maybe would not be bad to specifically say here "booster" in order to differentiate the loggers between the two.

File: services/distribution/src/test/java/app/coronawarn/server/services/distribution/assembly/component/BoosterNotificationStructureProviderTest.java [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+  DigitalCovidCertificateClient digitalCovidCertificateClient;
+
+  @MockBean
+  DigitalSigningCertificatesClient digitalSigningCertificatesClient;
+
+  @Autowired
+  BoosterNotificationStructureProvider underTest;
+
+  @Test
+  void should_create_correct_file_structure_for_booster_notification_business_rules() {
```
Comment (alexmoroianu) [inline services/distribution/src/test/java/app/coronawarn/server/services/distribution/assembly/component/BoosterNotificationStructureProviderTest.java:58]: I think @FelixRottler  wants to use camelCase in all the test methods. He also commented on my PRs about these. Maybe we should follow the same for consistency.

**Review by alexmoroianu — CHANGES_REQUESTED** [review]

File: services/distribution/src/test/java/app/coronawarn/server/services/distribution/assembly/component/BoosterNotificationStructureProviderTest.java (hunk 15 above)
Comment (ioangut) [reply services/distribution/src/test/java/app/coronawarn/server/services/distribution/assembly/component/BoosterNotificationStructureProviderTest.java:58]: I think we can keep this like it is because all the structure provider tests are all using this naming convention(maybe were not tests related to the structure providers)

Comment (FelixRottler) [reply services/distribution/src/test/java/app/coronawarn/server/services/distribution/assembly/component/BoosterNotificationStructureProviderTest.java:58]: I understand that it feels strange, but in the past we switched to camelCase. So I would agree with Alex here.

**Review by alexmoroianu — APPROVED** [review]

**Review by roesslerj — APPROVED** [review]
